echo-pubsub-google/src/main/java/com/netflix/spinnaker/echo/pubsub/google/GooglePubsubPublisher.java

```diff
@@ -0,0 +1,98 @@
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.echo.pubsub.google;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutureCallback;
+import com.google.api.core.ApiFutures;
+import com.google.api.gax.batching.BatchingSettings;
+import com.google.cloud.pubsub.v1.Publisher;
+import com.google.protobuf.ByteString;
+import com.google.pubsub.v1.ProjectTopicName;
+import com.google.pubsub.v1.PubsubMessage;
+import com.netflix.spinnaker.echo.config.GooglePubsubCredentialsProvider;
+import com.netflix.spinnaker.echo.config.GooglePubsubProperties.GooglePubsubPublisherConfig;
+import com.netflix.spinnaker.echo.model.pubsub.PubsubSystem;
+import com.netflix.spinnaker.echo.pubsub.model.PubsubPublisher;
+import java.io.IOException;
+import java.util.Map;
+import lombok.Data;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.threeten.bp.Duration;
+
+@Data
+@Slf4j
+public class GooglePubsubPublisher implements PubsubPublisher {
+
+  private final PubsubSystem pubsubSystem = PubsubSystem.GOOGLE;
+
+  private String topicName;
+
+  private Publisher publisher;
+
+  public static GooglePubsubPublisher buildPublisher(GooglePubsubPublisherConfig config) {
+    GooglePubsubPublisher publisher = new GooglePubsubPublisher();
+    ProjectTopicName fullName = ProjectTopicName.of(config.getProject(), config.getTopicName());
+    publisher.setTopicName(fullName.toString());
+
+    BatchingSettings batchingSettings = BatchingSettings.newBuilder()
+      .setElementCountThreshold(config.getBatchCountThreshold())
+      .setDelayThreshold(Duration.ofMillis(config.getDelayMillisecondsThreshold()))
+      .build();
+
+    try {
+      Publisher p = Publisher.newBuilder(fullName)
+        .setCredentialsProvider(new GooglePubsubCredentialsProvider(config.getJsonPath()))
+        .setBatchingSettings(batchingSettings)
+        .build();
+      publisher.setPublisher(p);
+    } catch (IOException ioe) {
+      log.error("Could not create Google Pubsub Publishers: {}", ioe);
+    }
+
+    return publisher;
+  }
+
+  public void publish(String jsonPayload, Map<String, String> attributes) {
+    PubsubMessage message = PubsubMessage.newBuilder()
+      .setData(ByteString.copyFromUtf8(jsonPayload))
+      .putAllAttributes(attributes)
+      .build();
+
+    log.debug("Publishing message on Google Pubsub topic {}", this.topicName);
+
+    ApiFuture<String> future = publisher.publish(message);
+    ApiFutures.addCallback(future, new PublishCallback(this.topicName));
+  }
+
+  @RequiredArgsConstructor
+  private class PublishCallback implements ApiFutureCallback<String> {
+
+    private final String topic;
+
+    @Override
+    public void onFailure(Throwable t) {
+      log.error("Could not publish message to Google Pubsub topic " + this.topic, t);
+    }
+
+    @Override
+    public void onSuccess(String result) {
+      log.debug("Successfully published message with ID {}", result);
```

Comment:
This is going to be a _ton_ of logs. I'd prefer to only log on failure.

Reply:
It's at the debug level, so it would only be a ton of logs if the operator wanted that. Do you guys run with debug enabled?

Reply:
Also, this is Google Pubsub specific.